### PR TITLE
 import referencesource Encoding foundation types.

### PR DESCRIPTION
**Pull request by atsushieno**
This brings Encoding and co., as well as basic encodings such as UTF*Encoding and ASCII (not including Latin1).
This change does not involve Windows-specific NLS-based encodings. We will be still using I18N assemblies until we make more grand design changes (Xamarin products premise those assemblies).

3rd shot, this should not really regress.

### CODE REVIEW

File: mcs/class/corlib/System.Text/EncodingHelper.cs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+	//
+	// Only internal, to be used by the class libraries: Unmarked and non-input-validating
+	//
+	internal static Encoding UTF8Unmarked {
+		get {
+			if (utf8EncodingWithoutMarkers == null) {
+				lock (lockobj){
+					if (utf8EncodingWithoutMarkers == null){
+						utf8EncodingWithoutMarkers = new UTF8Encoding (false, false);
+						typeof (Encoding).GetField ("m_isReadOnly", BindingFlags.NonPublic | BindingFlags.Instance).SetValue (utf8EncodingWithoutMarkers, false);
```
Comment (marek-safar) [inline mcs/class/corlib/System.Text/EncodingHelper.cs]: No reflection here to avoid linker issues

Comment (atsushieno) [inline mcs/class/corlib/System.Text/EncodingHelper.cs]: That's a good catch. Updated the fix.

**Comment by atsushieno** [comment]
build

**Comment by migueldeicaza** [comment]
So this means that we are still keeping the I18N assemblies around for the non-core encodings?   

These are the ones that are handled outside of mscorlib for now:

/mono/lib/mono/4.5/I18N.CJK.dll
/mono/lib/mono/4.5/I18N.MidEast.dll
/mono/lib/mono/4.5/I18N.Other.dll
/mono/lib/mono/4.5/I18N.Rare.dll
/mono/lib/mono/4.5/I18N.West.dll
/mono/lib/mono/4.5/I18N.dll

I did notice that Microsoft's encoding framework included a large table of encoding data.  I am not sure if this is in this patch.   Does this bring that encoding table, or is that a separate beast?

Xamarin products do use the I18N assemblies on the UI, to give users a chance to choose how much or little they want to include in their projects.   In total, the assemblies increase the DLL size by 1 megabyte.   Do you know how these compare to the table that Microsoft has?

**Comment by atsushieno** [comment]
Yes. That matches my concern and hence I didn't bring them (actally not easy to achive that in short effort). This patch doesn't bring those "NLP"-based encodings. EncodingNLS (which is NOT brought into mono by this change) handles those NLS resources, which are stored in external files on Windows system.
In these changes, I added call to "InvokeI18N" method in Microsoft's Encoding class so that it can still loads I18N encodings.

(also on trello: https://trello.com/c/d4SyuhY1/23-system-text-encoding)

coreclr doesn't handle non-Unicode encodings on Unix either.

**Pull request merged**